File: app/src/main/java/com/tvm/doctorcube/settings/FragmentEditDetails.java
package com.tvm.doctorcube.settings;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import com.bumptech.glide.Glide;
import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.tvm.doctorcube.R;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

import de.hdodenhof.circleimageview.CircleImageView;

public class FragmentEditDetails extends Fragment {

    private CircleImageView profileImageView;
    private ImageView editProfileImageBtn;
    private TextView userNameDisplay, applicationStatusText, neetScoreDisplay, universityDisplay, studyCountryDisplay;
    private TextView appliedDateText, lastUpdatedText, lastCallDateText, universityNameText;
    private TextInputEditText nameEditText, emailEditText, mobileEditText, countryEditText, stateEditText, cityEditText;
    private TextInputEditText neetScoreEditText, studyCountryEditText;
    private CheckBox hasNeetScoreCheckbox, hasPassportCheckbox;
    private Button editModeToggleBtn, cancelBtn, saveBtn, callBtn, whatsappBtn, viewUniversityDetailsBtn;
    private View actionButtonsLayout;

    private FirebaseAuth mAuth;
    private DocumentReference userRef;
    private StorageReference storageRef;
    private boolean isEditMode = false;
    private Uri imageUri;
    private ActivityResultLauncher<Intent> imagePickerLauncher;
    private SimpleDateFormat dateFormat = new SimpleDateFormat("dd/MM/yy", Locale.getDefault());

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_edit_details, container, false);
        initializeViews(view);
        initializeFirebase();
        setupImagePicker();
        loadUserData();
        setupClickListeners();
        return view;
    }

    private void initializeViews(View view) {
        profileImageView = view.findViewById(R.id.profileImageView);
        editProfileImageBtn = view.findViewById(R.id.editProfileImageBtn);
        userNameDisplay = view.findViewById(R.id.userNameDisplay);
        applicationStatusText = view.findViewById(R.id.applicationStatusText);
        neetScoreDisplay = view.findViewById(R.id.neetScoreDisplay);
        universityDisplay = view.findViewById(R.id.universityDisplay);
        studyCountryDisplay = view.findViewById(R.id.studyCountryDisplay);
        appliedDateText = view.findViewById(R.id.appliedDateText);
        lastUpdatedText = view.findViewById(R.id.lastUpdatedText);
        lastCallDateText = view.findViewById(R.id.lastCallDateText);
        universityNameText = view.findViewById(R.id.universityNameText);
        nameEditText = view.findViewById(R.id.nameEditText);
        emailEditText = view.findViewById(R.id.emailEditText);
        mobileEditText = view.findViewById(R.id.mobileEditText);
        countryEditText = view.findViewById(R.id.countryEditText);
        stateEditText = view.findViewById(R.id.stateEditText);
        cityEditText = view.findViewById(R.id.cityEditText);
        neetScoreEditText = view.findViewById(R.id.neetScoreEditText);
        studyCountryEditText = view.findViewById(R.id.studyCountryEditText);
        hasNeetScoreCheckbox = view.findViewById(R.id.hasNeetScoreCheckbox);
        hasPassportCheckbox = view.findViewById(R.id.hasPassportCheckbox);
        editModeToggleBtn = view.findViewById(R.id.editModeToggleBtn);
        cancelBtn = view.findViewById(R.id.cancelBtn);
        saveBtn = view.findViewById(R.id.saveBtn);
        callBtn = view.findViewById(R.id.callBtn);
        whatsappBtn = view.findViewById(R.id.whatsappBtn);
        viewUniversityDetailsBtn = view.findViewById(R.id.viewUniversityDetailsBtn);
        actionButtonsLayout = view.findViewById(R.id.actionButtonsLayout);
    }

    private void initializeFirebase() {
        mAuth = FirebaseAuth.getInstance();
        String userId = mAuth.getCurrentUser() != null ? mAuth.getCurrentUser().getUid() : "";
        if (userId.isEmpty()) {
            Toast.makeText(getContext(), "User not authenticated", Toast.LENGTH_SHORT).show();
            setDefaultUI();
            return;
        }
        userRef = FirebaseFirestore.getInstance().collection("app_submissions").document(userId);
        storageRef = FirebaseStorage.getInstance().getReference("profile_images").child(userId + "/profile.jpg");
    }

    private void setupImagePicker() {
        imagePickerLauncher = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(),
                result -> {
                    if (result.getResultCode() == Activity.RESULT_OK && result.getData() != null) {
                        imageUri = result.getData().getData();
                        Glide.with(this).load(imageUri).into(profileImageView);
                    }
                });
    }

    private void loadUserData() {
        if (userRef == null) {
            setDefaultUI();
            return;
        }

        userRef.get().addOnSuccessListener(documentSnapshot -> {
            if (!isAdded()) return;

            if (documentSnapshot.exists()) {
                // Set default values if fields don't exist
                String name = documentSnapshot.contains("name") ? documentSnapshot.getString("name") : "Unknown User";
                String email = documentSnapshot.contains("email") ? documentSnapshot.getString("name") : "No Email";
                String mobile = documentSnapshot.contains("mobile") ? documentSnapshot.getString("mobile") : "No Mobile";
                String country = documentSnapshot.contains("country") ? documentSnapshot.getString("country") : "Unknown";
                String state = documentSnapshot.contains("state") ? documentSnapshot.getString("state") : "Unknown";
                String city = documentSnapshot.contains("city") ? documentSnapshot.getString("city") : "Unknown";
                String neetScore = documentSnapshot.contains("neetScore") ? documentSnapshot.getString("neetScore") : "0";
                String studyCountry = documentSnapshot.contains("studyCountry") ? documentSnapshot.getString("studyCountry") : "Unknown";
                String universityName = documentSnapshot.contains("universityName") ? documentSnapshot.getString("universityName") : "Unknown University";
                String lastCallDate = documentSnapshot.contains("lastCallDate") ? documentSnapshot.getString("lastCallDate") : "Not Called Yet";
                Boolean hasNeetScore = documentSnapshot.contains("hasNeetScore") ? documentSnapshot.getBoolean("hasNeetScore") : false;
                Boolean hasPassport = documentSnapshot.contains("hasPassport") ? documentSnapshot.getBoolean("hasPassport") : false;
                Boolean isAdmitted = documentSnapshot.contains("isAdmitted") ? documentSnapshot.getBoolean("isAdmitted") : false;
                Boolean isApplied = documentSnapshot.contains("isApplied") ? documentSnapshot.getBoolean("isApplied") : false;

                // Handle dates
                String appliedDate = documentSnapshot.contains("appliedDate") ?
                        formatTimestamp(documentSnapshot.getTimestamp("appliedDate")) : "N/A";
                String lastUpdated = documentSnapshot.contains("lastUpdatedDate") ?
                        formatStringDate(documentSnapshot.getString("lastUpdatedDate")) : dateFormat.format(new Date());

                // Determine application status
                String status = isAdmitted ? "Admitted" : isApplied ? "Application Submitted" : lastCallDate.equals("Not Called Yet") ? "Under Review" : "Applied";

                // Update UI
                userNameDisplay.setText(name);
                applicationStatusText.setText(status);
                neetScoreDisplay.setText(neetScore);
                universityDisplay.setText(universityName);
                studyCountryDisplay.setText(studyCountry);
                appliedDateText.setText(appliedDate);
                lastUpdatedText.setText(lastUpdated);
                lastCallDateText.setText(lastCallDate);
                universityNameText.setText(universityName);
                nameEditText.setText(name);
                emailEditText.setText(email);
                mobileEditText.setText(mobile);
                countryEditText.setText(country);
                stateEditText.setText(state);
                cityEditText.setText(city);
                neetScoreEditText.setText(neetScore);
                studyCountryEditText.setText(studyCountry);
                hasNeetScoreCheckbox.setChecked(hasNeetScore);
                hasPassportCheckbox.setChecked(hasPassport);

                // Update status badge color
                int statusColor = isAdmitted ? R.color.medical_accent_green :
                        status.equals("Under Review") ? R.color.status_warning : R.color.medical_accent_green;
                applicationStatusText.getRootView().setBackgroundTintList(getResources().getColorStateList(statusColor, getContext().getTheme()));

                // Load profile image
                loadProfileImage();
            } else {
                Toast.makeText(getContext(), "No user data found", Toast.LENGTH_SHORT).show();
                setDefaultUI();
            }
        }).addOnFailureListener(e -> {
            if (!isAdded()) return;
            if (e instanceof FirebaseFirestoreException && ((FirebaseFirestoreException) e).getCode() == FirebaseFirestoreException.Code.PERMISSION_DENIED) {
                Toast.makeText(getContext(), "Permission denied. Please check your authentication status.", Toast.LENGTH_LONG).show();
            } else {
                Toast.makeText(getContext(), "Failed to load data: " + e.getMessage(), Toast.LENGTH_SHORT).show();
            }
            setDefaultUI();
        });
    }

    private void setDefaultUI() {
        String defaultValue = "N/A";
        userNameDisplay.setText("Unknown User");
        applicationStatusText.setText("Under Review");
        neetScoreDisplay.setText("0");
        universityDisplay.setText("Unknown University");
        studyCountryDisplay.setText("Unknown");
        appliedDateText.setText(defaultValue);
        lastUpdatedText.setText(dateFormat.format(new Date()));
        lastCallDateText.setText("Not Called Yet");
        universityNameText.setText("Unknown University");
        nameEditText.setText("Unknown User");
        emailEditText.setText("No Email");
        mobileEditText.setText("No Mobile");
        countryEditText.setText("Unknown");
        stateEditText.setText("Unknown");
        cityEditText.setText("Unknown");
        neetScoreEditText.setText("0");
        studyCountryEditText.setText("Unknown");
        hasNeetScoreCheckbox.setChecked(false);
        hasPassportCheckbox.setChecked(false);
        applicationStatusText.getRootView().setBackgroundTintList(getResources().getColorStateList(R.color.status_warning, getContext().getTheme()));
        loadProfileImage();
    }

    private String formatTimestamp(com.google.firebase.Timestamp timestamp) {
        if (timestamp == null) return "N/A";
        try {
            Date date = timestamp.toDate();
            return dateFormat.format(date);
        } catch (Exception e) {
            return "N/A";
        }
    }

    private String formatStringDate(String dateStr) {
        if (dateStr == null || dateStr.isEmpty()) return "N/A";
        try {
            Date date = dateFormat.parse(dateStr);
            return dateFormat.format(date);
        } catch (ParseException e) {
            return "N/A";
        }
    }

    private void loadProfileImage() {
        storageRef.getDownloadUrl().addOnSuccessListener(uri -> {
            if (isAdded()) {
                Glide.with(FragmentEditDetails.this).load(uri).placeholder(R.drawable.ic_profile).into(profileImageView);
            }
        }).addOnFailureListener(e -> {
            if (isAdded()) {
                profileImageView.setImageResource(R.drawable.ic_profile);
            }
        });
    }

    private void setupClickListeners() {
        editModeToggleBtn.setOnClickListener(v -> toggleEditMode());
        cancelBtn.setOnClickListener(v -> toggleEditMode());
        saveBtn.setOnClickListener(v -> saveChanges());
        editProfileImageBtn.setOnClickListener(v -> pickImage());
        callBtn.setOnClickListener(v -> initiateCall());
        whatsappBtn.setOnClickListener(v -> initiateWhatsApp());
        viewUniversityDetailsBtn.setOnClickListener(v -> viewUniversityDetails());
    }

    private void toggleEditMode() {
        isEditMode = !isEditMode;
        editModeToggleBtn.setText(isEditMode ? "Done" : "Edit");
        actionButtonsLayout.setVisibility(isEditMode ? View.VISIBLE : View.GONE);
        editProfileImageBtn.setVisibility(isEditMode ? View.VISIBLE : View.GONE);
        setFieldsEditable(isEditMode);
    }

    private void setFieldsEditable(boolean editable) {
        nameEditText.setEnabled(editable);
        emailEditText.setEnabled(editable);
        mobileEditText.setEnabled(editable);
        countryEditText.setEnabled(editable);
        stateEditText.setEnabled(editable);
        cityEditText.setEnabled(editable);
        neetScoreEditText.setEnabled(editable);
        studyCountryEditText.setEnabled(editable);
        hasNeetScoreCheckbox.setEnabled(editable);
        hasPassportCheckbox.setEnabled(editable);
    }

    private void pickImage() {
        Intent intent = new Intent(Intent.ACTION_PICK);
        intent.setType("image/*");
        imagePickerLauncher.launch(intent);
    }

    private void saveChanges() {
        if (userRef == null) {
            Toast.makeText(getContext(), "Cannot save: User not authenticated", Toast.LENGTH_SHORT).show();
            return;
        }

        Map<String, Object> updates = new HashMap<>();
        updates.put("name", nameEditText.getText().toString().trim());
        updates.put("email", emailEditText.getText().toString().trim());
        updates.put("mobile", mobileEditText.getText().toString().trim());
        updates.put("country", countryEditText.getText().toString().trim());
        updates.put("state", stateEditText.getText().toString().trim());
        updates.put("city", cityEditText.getText().toString().trim());
        updates.put("neetScore", neetScoreEditText.getText().toString().trim());
        updates.put("studyCountry", studyCountryEditText.getText().toString().trim());
        updates.put("hasNeetScore", hasNeetScoreCheckbox.isChecked());
        updates.put("hasPassport", hasPassportCheckbox.isChecked());
        updates.put("lastUpdatedDate", dateFormat.format(new Date()));

        userRef.set(updates).addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                Toast.makeText(getContext(), "Profile updated successfully", Toast.LENGTH_SHORT).show();
                toggleEditMode();
            } else {
                String errorMsg = task.getException() != null ? task.getException().getMessage() : "Unknown error";
                if (task.getException() instanceof FirebaseFirestoreException &&
                        ((FirebaseFirestoreException) task.getException()).getCode() == FirebaseFirestoreException.Code.PERMISSION_DENIED) {
                    Toast.makeText(getContext(), "Permission denied. Please check your authentication status.", Toast.LENGTH_LONG).show();
                } else {
                    Toast.makeText(getContext(), "Failed to update profile: " + errorMsg, Toast.LENGTH_SHORT).show();
                }
            }
        });

        if (imageUri != null) {
            storageRef.putFile(imageUri).addOnCompleteListener(task -> {
                if (task.isSuccessful()) {
                    loadProfileImage();
                } else {
                    String errorMsg = task.getException() != null ? task.getException().getMessage() : "Unknown error";
                    Toast.makeText(getContext(), "Failed to upload image: " + errorMsg, Toast.LENGTH_SHORT).show();
                }
            });
        }
    }

    private void initiateCall() {
        Intent callIntent = new Intent(Intent.ACTION_DIAL);
        callIntent.setData(Uri.parse("tel:+1234567890")); // Replace with actual number
        startActivity(callIntent);

        if (userRef != null) {
            userRef.update("lastCallDate", dateFormat.format(new Date())).addOnFailureListener(e -> {
                Toast.makeText(getContext(), "Failed to update call date: " + e.getMessage(), Toast.LENGTH_SHORT).show();
            });
        }
    }

    private void initiateWhatsApp() {
        String phoneNumber = "+1234567890"; // Replace with actual number
        String url = "https://wa.me/" + phoneNumber;
        Intent whatsappIntent = new Intent(Intent.ACTION_VIEW);
        whatsappIntent.setData(Uri.parse(url));
        startActivity(whatsappIntent);
    }

    private void viewUniversityDetails() {
        Toast.makeText(getContext(), "Navigating to university details", Toast.LENGTH_SHORT).show();
        // Implement navigation to university details screen
    }
}

File: app/src/main/java/com/tvm/doctorcube/settings/FragmentEditDetails.kt
package com.tvm.doctorcube.settings

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import com.bumptech.glide.Glide
import com.google.firebase.Timestamp
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.FirebaseFirestoreException
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference
import com.tvm.doctorcube.R
import com.tvm.doctorcube.databinding.FragmentEditDetailsBinding
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class FragmentEditDetails : Fragment() {
    private var _binding: FragmentEditDetailsBinding? = null
    private val binding get() = _binding!!

    private var mAuth: FirebaseAuth? = null
    private var userRef: DocumentReference? = null
    private var storageRef: StorageReference? = null
    private var documentListener: ListenerRegistration? = null

    private var isEditMode = false
    private var imageUri: Uri? = null
    private var cachedUserData: DocumentSnapshot? = null
    private var hasUnsavedChanges = false
    private var lastKnownData = mutableMapOf<String, Any?>()

    private val dateFormat = SimpleDateFormat("dd/MM/yy", Locale.getDefault())

    private val imagePickerLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode == Activity.RESULT_OK && result.data != null) {
            imageUri = result.data?.data
            imageUri?.let {
                Glide.with(this).load(it).into(binding.profileImageView)
                hasUnsavedChanges = true
            }
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentEditDetailsBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initializeFirebase()
        setupRealtimeListener()
        setupClickListeners()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        documentListener?.remove()
        _binding = null
    }

    private fun initializeFirebase() {
        mAuth = FirebaseAuth.getInstance()
        val userId = mAuth?.currentUser?.uid ?: run {
            showToast("User not authenticated")
            setDefaultUI()
            return
        }
        userRef = FirebaseFirestore.getInstance().collection("app_submissions").document(userId)
        storageRef = FirebaseStorage.getInstance().getReference("profile_images")
            .child("$userId/profile.jpg")
    }

    // Use real-time listener instead of manual refresh
    private fun setupRealtimeListener() {
        userRef?.let { ref ->
            documentListener = ref.addSnapshotListener { documentSnapshot, error ->
                if (!isAdded) return@addSnapshotListener

                error?.let {
                    handleFirebaseError(it)
                    return@addSnapshotListener
                }

                documentSnapshot?.let { snapshot ->
                    if (snapshot.exists()) {
                        cachedUserData = snapshot
                        updateUIFromCache()
                        cacheCurrentData(snapshot)
                    } else {
                        showToast("No user data found")
                        setDefaultUI()
                    }
                }
            }
        }
    }

    private fun cacheCurrentData(snapshot: DocumentSnapshot) {
        lastKnownData.clear()
        snapshot.data?.let { data ->
            lastKnownData.putAll(data)
        }
    }

    private fun updateUIFromCache() {
        cachedUserData?.let { snapshot ->
            val userData = UserData.fromSnapshot(snapshot)
            populateUI(userData)
            loadProfileImageOnce()
        }
    }

    private fun populateUI(userData: UserData) {
        with(binding) {
            // Header Section
            setTextAndVisibility(userNameDisplay, userData.name, "Unknown User")
            applicationStatusText.text = userData.status
            setTextAndVisibility(neetScoreDisplay, userData.neetScore, "0")
            setTextAndVisibility(universityDisplay, userData.universityName, "Unknown University")
            setTextAndVisibility(studyCountryDisplay, userData.studyCountry, "Unknown")

            // Application Status Section
            appliedDateText.text = userData.appliedDate
            lastUpdatedText.text = userData.lastUpdated
            setTextAndVisibility(lastCallDateText, userData.lastCallDate, hideIfDefault = "Not Called Yet")
            setTextAndVisibility(universityNameText, userData.universityName, "Unknown University")

            // Edit Profile Section - only populate if in edit mode or first load
            if (!isEditMode || !hasUnsavedChanges) {
                populateEditFields(userData)
            }

            // Set visibility based on edit mode and data availability
            updateFieldVisibility(userData)

            // Set status color
            val statusColor = when {
                userData.isAdmitted -> R.color.medical_accent_green
                userData.status == "Under Review" -> R.color.status_warning
                else -> R.color.medical_accent_green
            }
            applicationStatusText.backgroundTintList = ContextCompat.getColorStateList(requireContext(), statusColor)
        }
    }

    private fun populateEditFields(userData: UserData) {
        with(binding) {
            nameEditText.setText(userData.name)
            emailEditText.setText(userData.email)
            mobileEditText.setText(userData.mobile)
            countryEditText.setText(userData.country)
            stateEditText.setText(userData.state)
            cityEditText.setText(userData.city)
            neetScoreEditText.setText(userData.neetScore)
            studyCountryEditText.setText(userData.studyCountry)
            hasNeetScoreCheckbox.isChecked = userData.hasNeetScore
            hasPassportCheckbox.isChecked = userData.hasPassport
        }
    }

    private fun updateFieldVisibility(userData: UserData) {
        with(binding) {
            if (isEditMode) {
                // Show all fields in edit mode
                listOf(
                    nameInputLayout, emailInputLayout, mobileInputLayout,
                    countryInputLayout, stateInputLayout, cityInputLayout,
                    neetScoreInputLayout, studyCountryInputLayout,
                    hasNeetScoreCheckbox, hasPassportCheckbox
                ).forEach { it.visibility = View.VISIBLE }
            } else {
                // Hide empty fields in view mode
                nameInputLayout.visibility = if (userData.name.isNullOrBlank()) View.GONE else View.VISIBLE
                emailInputLayout.visibility = if (userData.email.isNullOrBlank()) View.GONE else View.VISIBLE
                mobileInputLayout.visibility = if (userData.mobile.isNullOrBlank()) View.GONE else View.VISIBLE
                countryInputLayout.visibility = if (userData.country.isNullOrBlank()) View.GONE else View.VISIBLE
                stateInputLayout.visibility = if (userData.state.isNullOrBlank()) View.GONE else View.VISIBLE
                cityInputLayout.visibility = if (userData.city.isNullOrBlank()) View.GONE else View.VISIBLE
                neetScoreInputLayout.visibility = if (userData.neetScore.isNullOrBlank()) View.GONE else View.VISIBLE
                studyCountryInputLayout.visibility = if (userData.studyCountry.isNullOrBlank()) View.GONE else View.VISIBLE
                hasNeetScoreCheckbox.visibility = if (!userData.hasNeetScore) View.GONE else View.VISIBLE
                hasPassportCheckbox.visibility = if (!userData.hasPassport) View.GONE else View.VISIBLE
            }
        }
    }

    private fun setTextAndVisibility(view: View, value: String?, defaultValue: String = "", hideIfDefault: String? = null) {
        when (view) {
            is androidx.appcompat.widget.AppCompatTextView -> {
                view.text = value ?: defaultValue
                val shouldHide = value.isNullOrBlank() || (hideIfDefault != null && value == hideIfDefault)
                (view.parent as? View)?.visibility = if (shouldHide) View.GONE else View.VISIBLE
            }
        }
    }

    private var profileImageLoaded = false
    private fun loadProfileImageOnce() {
        if (profileImageLoaded) return

        storageRef?.downloadUrl?.addOnSuccessListener { uri ->
            if (isAdded) {
                Glide.with(this@FragmentEditDetails)
                    .load(uri)
                    .placeholder(R.drawable.ic_profile)
                    .into(binding.profileImageView)
                profileImageLoaded = true
            }
        }?.addOnFailureListener {
            if (isAdded) {
                binding.profileImageView.setImageResource(R.drawable.ic_profile)
                profileImageLoaded = true
            }
        }
    }

    private fun setDefaultUI() {
        val defaultUserData = UserData()
        populateUI(defaultUserData)
        binding.applicationStatusText.backgroundTintList =
            ContextCompat.getColorStateList(requireContext(), R.color.status_warning)
        loadProfileImageOnce()
    }

    private fun setupClickListeners() {
        with(binding) {
            editModeToggleBtn.setOnClickListener { toggleEditMode() }
            cancelBtn.setOnClickListener { cancelChanges() }
            saveBtn.setOnClickListener { saveChanges() }
            editProfileImageBtn.setOnClickListener { pickImage() }
            callBtn.setOnClickListener { initiateCall() }
            whatsappBtn.setOnClickListener { initiateWhatsApp() }
        }
    }

    private fun toggleEditMode() {
        isEditMode = !isEditMode
        hasUnsavedChanges = false

        with(binding) {
            editModeToggleBtn.text = if (isEditMode) "Done" else "Edit"
            actionButtonsLayout.visibility = if (isEditMode) View.VISIBLE else View.GONE
            editProfileImageBtn.visibility = if (isEditMode) View.VISIBLE else View.GONE
        }

        setFieldsEditable(isEditMode)
        cachedUserData?.let {
            val userData = UserData.fromSnapshot(it)
            updateFieldVisibility(userData)
        }
    }

    private fun cancelChanges() {
        hasUnsavedChanges = false
        imageUri = null
        cachedUserData?.let {
            val userData = UserData.fromSnapshot(it)
            populateEditFields(userData)
        }
        toggleEditMode()
    }

    private fun setFieldsEditable(editable: Boolean) {
        with(binding) {
            listOf(
                nameEditText, emailEditText, mobileEditText,
                countryEditText, stateEditText, cityEditText,
                neetScoreEditText, studyCountryEditText,
                hasNeetScoreCheckbox, hasPassportCheckbox
            ).forEach { it.isEnabled = editable }
        }
    }

    private fun pickImage() {
        val intent = Intent(Intent.ACTION_PICK).apply { type = "image/*" }
        imagePickerLauncher.launch(intent)
    }

    private fun saveChanges() {
        userRef ?: return

        val newData = collectFormData()
        val changedFields = getChangedFields(newData)

        if (changedFields.isEmpty() && imageUri == null) {
            showToast("No changes to save")
            toggleEditMode()
            return
        }

        lifecycleScope.launch {
            try {
                // Only update changed fields
                if (changedFields.isNotEmpty()) {
                    changedFields["lastUpdatedDate"] = dateFormat.format(Date())
                    userRef?.update(changedFields)?.await()
                }

                // Upload image if changed
                imageUri?.let { uri ->
                    storageRef?.putFile(uri)?.await()
                    profileImageLoaded = false // Force reload
                    imageUri = null
                }

                if (isAdded) {
                    showToast("Profile updated successfully")
                    hasUnsavedChanges = false
                    toggleEditMode()
                }
            } catch (e: Exception) {
                if (isAdded) {
                    handleFirebaseError(e)
                }
            }
        }
    }

    private fun collectFormData(): Map<String, Any?> {
        return with(binding) {
            mapOf(
                "name" to nameEditText.text.toString().trim(),
                "email" to emailEditText.text.toString().trim(),
                "mobile" to mobileEditText.text.toString().trim(),
                "country" to countryEditText.text.toString().trim(),
                "state" to stateEditText.text.toString().trim(),
                "city" to cityEditText.text.toString().trim(),
                "neetScore" to neetScoreEditText.text.toString().trim(),
                "studyCountry" to studyCountryEditText.text.toString().trim(),
                "hasNeetScore" to hasNeetScoreCheckbox.isChecked,
                "hasPassport" to hasPassportCheckbox.isChecked
            )
        }
    }

    private fun getChangedFields(newData: Map<String, Any?>): MutableMap<String, Any?> {
        val changes = mutableMapOf<String, Any?>()

        newData.forEach { (key, newValue) ->
            val oldValue = lastKnownData[key]
            if (oldValue != newValue) {
                changes[key] = newValue
            }
        }

        return changes
    }

    private fun initiateCall() {
        val callIntent = Intent(Intent.ACTION_DIAL).apply {
            data = Uri.parse("tel:${getString(R.string.whatsapp_number)}")
        }
        startActivity(callIntent)

        // Only update if call date has actually changed
        val today = dateFormat.format(Date())
        val lastCallDate = lastKnownData["lastCallDate"] as? String

        if (lastCallDate != today) {
            lifecycleScope.launch {
                try {
                    userRef?.update("lastCallDate", today)?.await()
                } catch (e: Exception) {
                    if (isAdded) {
                        showToast("Failed to update call date: ${e.message}")
                    }
                }
            }
        }
    }

    private fun initiateWhatsApp() {
        val phoneNumber = getString(R.string.whatsapp_number)
        val url = "https://wa.me/$phoneNumber"
        val whatsappIntent = Intent(Intent.ACTION_VIEW).apply {
            data = Uri.parse(url)
        }
        startActivity(whatsappIntent)
    }

    private fun handleFirebaseError(exception: Exception) {
        val message = when (exception) {
            is FirebaseFirestoreException -> {
                if (exception.code == FirebaseFirestoreException.Code.PERMISSION_DENIED) {
                    "Permission denied. Please check your authentication status."
                } else {
                    "Database error: ${exception.message}"
                }
            }
            else -> "Failed to process request: ${exception.message}"
        }
        showToast(message)
    }

    private fun showToast(message: String) {
        if (isAdded) {
            Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
        }
    }

    // Data class to structure user data
    private data class UserData(
        val name: String? = null,
        val email: String? = null,
        val mobile: String? = null,
        val country: String? = null,
        val state: String? = null,
        val city: String? = null,
        val neetScore: String? = null,
        val studyCountry: String? = null,
        val universityName: String? = null,
        val lastCallDate: String = "Not Called Yet",
        val hasNeetScore: Boolean = false,
        val hasPassport: Boolean = false,
        val isAdmitted: Boolean = false,
        val isApplied: Boolean = false,
        val appliedDate: String = "N/A",
        val lastUpdated: String = SimpleDateFormat("dd/MM/yy", Locale.getDefault()).format(Date())
    ) {
        val status: String
            get() = when {
                isAdmitted -> "Admitted"
                isApplied -> "Application Submitted"
                lastCallDate == "Not Called Yet" -> "Under Review"
                else -> "Applied"
            }

        companion object {
            fun fromSnapshot(snapshot: DocumentSnapshot): UserData {
                val dateFormat = SimpleDateFormat("dd/MM/yy", Locale.getDefault())

                return UserData(
                    name = snapshot.getString("name"),
                    email = snapshot.getString("email"),
                    mobile = snapshot.getString("mobile"),
                    country = snapshot.getString("country"),
                    state = snapshot.getString("state"),
                    city = snapshot.getString("city"),
                    neetScore = snapshot.getString("neetScore"),
                    studyCountry = snapshot.getString("studyCountry"),
                    universityName = snapshot.getString("universityName"),
                    lastCallDate = snapshot.getString("lastCallDate") ?: "Not Called Yet",
                    hasNeetScore = snapshot.getBoolean("hasNeetScore") ?: false,
                    hasPassport = snapshot.getBoolean("hasPassport") ?: false,
                    isAdmitted = snapshot.getBoolean("isAdmitted") ?: false,
                    isApplied = snapshot.getBoolean("isApplied") ?: false,
                    appliedDate = snapshot.getTimestamp("timestamp")?.let {
                        try { dateFormat.format(it.toDate()) } catch (e: Exception) { "N/A" }
                    } ?: "N/A",
                    lastUpdated = snapshot.getString("lastUpdatedDate")?.let {
                        try {
                            dateFormat.parse(it)?.let { dateFormat.format(it) } ?: dateFormat.format(Date())
                        } catch (e: ParseException) {
                            dateFormat.format(Date())
                        }
                    } ?: dateFormat.format(Date())
                )
            }
        }
    }
}